aten/src/ATen/native/attn.cpp

```diff
@@ -0,0 +1,86 @@
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+std::tuple<Tensor, Tensor> attn(
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v) {
+
+  TORCH_CHECK(q.dim() == 2 && k.dim() == 2 && v.dim() == 2, "All of q,k,v need to be 2D tensor");
+  TORCH_CHECK(q.size(1) == k.size(1), "q and k must have must same size for 2nd dim");
+  TORCH_CHECK(k.size(0) == v.size(0), "k and v must have must same size for 1st dim");
+
+  Tensor x = at::matmul(q, k.t());
+  Tensor a = at::tanh(x);
+  Tensor o = at::matmul(a, v);
+
+  return std::make_tuple(o, a);
+}
+
+std::tuple<Tensor, Tensor, Tensor> attn_backward(
+    const Tensor& do_ext,
+    const Tensor& da_ext,
+    const Tensor& q,
+    const Tensor& k,
+    const Tensor& v,
+    const Tensor& a) {
+
+  Tensor dq, dk, dv;
+
+  // Alternatively, this check can be expressed in derivatives.yaml as
+  // - q, k, v: "grads[0].defined() || grads[1].defined() ? attn_backward(grads[0], grads[1], q, k, v, a) : std::tuple<Tensor, Tensor, Tensor>()"
```

Comment:
@albanD, I presume it's also preferred to write this in the derivatives.yaml file? since such corner case will bailout earlier in the stack?

Reply:
This check is not actually needed.
The execution engine, for a particular call to backward(), will skip this Node if no gradients flow to it.

Reply:
hmm, I am seeing a lot of examples that explicitly check whether grad is defined, e.g.

```
- name: prelu(Tensor self, Tensor weight) -> Tensor
  self, weight: "grad.defined() ? prelu_backward(grad, self, weight) : std::tuple<Tensor, Tensor>()"
```

I thought this check is used to bypass the backward compute, but allow the empty gradient to further back-propagate?

Reply:
I would guess unnecessary code :D

Reply:
I tried skipping the check on whether grad is defined, and gradcheck is failing with following message. 
I think it's still needed to handle undefined grad case explicitly. 

  File "/data/home/bahuang/repos/pytorch/torch/autograd/gradcheck.py", line 958, in check_undefined_grad_support
    raise GradcheckError((
torch.autograd.gradcheck.GradcheckError: **Expected backward function to handle undefined output grads. Please look at "Notes about undefined output gradients**" in "tools/autograd/derivatives.yaml"